compiler/src/yul/runtime/functions/math.rs

```diff
@@ -38,7 +57,47 @@ pub fn checked_sub_fns() -> Vec<yul::Statement> {
 
 // Return all math runtime functions
 pub fn all() -> Vec<yul::Statement> {
-    [checked_add_fns(), checked_sub_fns()].concat()
+    [checked_add_fns(), checked_mul_fns(), checked_sub_fns()].concat()
+}
+
+
+fn checked_mul_unsigned(size: Integer) -> yul::Statement {
+    if size.is_signed() {
+        panic!("Expected unsigned integer")
+    }
+    let fn_name = names::checked_mul(&size);
+    let max_value = get_max(size);
+
+    function_definition! {
+        function [fn_name](val1, val2) -> product {
+            // overflow, if val1 != 0 and val2 > (max_value / val1)
+            (if (and((iszero((iszero(val1)))), (gt(val2, (div([max_value], val1)))))) { (revert(0, 0)) })
+            (product := mul(val1, val2))
+        }
+    }
+}
+
+fn checked_mul_signed(size: Integer) -> yul::Statement {
+    if !size.is_signed() {
+        panic!("Expected signed integer")
+    }
+    let fn_name = names::checked_mul(&size);
+    let (min_value, max_value) = get_min_max(size.clone());
+
+    function_definition! {
+        function [fn_name](val1, val2) -> product {
+            // overflow, if val1 > 0, val2 > 0 and val1 > (max_value / val2)
+            (if (and((and((sgt(val1, 0)), (sgt(val2, 0)))), (gt(val1, (div([max_value.clone()], val2)))))) { (revert(0, 0)) })
+            // underflow, if val1 > 0, val2 < 0 and val2 < (min_value / val1)
+            (if (and((and((sgt(val1, 0)), (slt(val2, 0)))), (slt(val2, (sdiv([min_value.clone()], val1)))))) { (revert(0, 0)) })
+            // underflow, if val1 < 0, val2 > 0 and val1 < (min_value / val2)
+            // THIS IS THE CHECK THAT CAUSES THE REVERT
+            (if (and((and((slt(val1, 0)), (sgt(val2, 0)))), (slt(val1, (sdiv([min_value], val2)))))) { (revert(0, 0)) })
```

Comment:
This is the check that causes us to revert. We detect an underflow. The check is 100 % identical to solidity. Only the variables are renamed but I double checked the generated yul functions and logic is exactly the same what my `solc 0.6.7+commit.b8d736ae.Linux.g++` generates.